File: src/android/TheoremReachPlugin.java
/********* TheoremReach.m Cordova Plugin Implementation *******/

#import <Cordova/CDV.h>
#import <TheoremReachSDK/TheoremReach.h>

NSString *onRewardCallback;
NSString *onRewardCenterOpenedCallback;
NSString *onRewardCenterClosedCallback;
NSString *theoremreachSurveyAvailableCallback;

@interface TheoremReachPlugin : CDVPlugin <TheoremReachRewardDelegate, TheoremReachSurveyDelegate> {
  // Member variables go here.
}

- (void)initWithApiKeyAndUserId:(CDVInvokedUrlCommand*)command;
- (void)enableDebugMode:(CDVInvokedUrlCommand*)command;
- (void)isSurveyAvailable:(CDVInvokedUrlCommand*)command;
- (void)showRewardCenter:(CDVInvokedUrlCommand*)command;
- (void)setRewardCallback:(CDVInvokedUrlCommand*)command;
- (void)setRewardCenterOpenedCallback:(CDVInvokedUrlCommand*)command;
- (void)setRewardCenterClosedCallback:(CDVInvokedUrlCommand*)command;
- (void)setTheoremreachSurveyAvailableCallback:(CDVInvokedUrlCommand*)command;
@end

@implementation TheoremReachPlugin

- (void)initWithApiKeyAndUserId:(CDVInvokedUrlCommand*)command
{
    NSString* apiKey = [command.arguments objectAtIndex:0];
    NSString* userId = [command.arguments objectAtIndex:1];

    [self.commandDelegate runInBackground:^{
        [TheoremReach initWithApiKey:apiKey userId:userId];
        
        [[TheoremReach getInstance] setRewardListenerDelegate:self];
        [[TheoremReach getInstance] setSurveyListenerDelegate:self];
    }];
}

- (void)enableDebugMode:(CDVInvokedUrlCommand*)command
{
    BOOL debugMode = [command.arguments objectAtIndex:0];

    [[TheoremReach getInstance] enableDebugMode:debugMode];
}

- (void)isSurveyAvailable:(CDVInvokedUrlCommand*)command
{
    [self.commandDelegate runInBackground:^{
        CDVPluginResult* pluginResult = nil;
        BOOL result = [TheoremReach getInstance].isSurveyAvailable;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:result];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)showRewardCenter:(CDVInvokedUrlCommand*)command
{
    NSLog(@"TheoremReachPlugin RewardCenter Opening");
    [TheoremReach showRewardCenter];
}

- (void)setRewardCallback:(CDVInvokedUrlCommand*)command
{
    onRewardCallback = command.callbackId;
}

- (void)setRewardCenterOpenedCallback:(CDVInvokedUrlCommand*)command
{
    onRewardCenterOpenedCallback = command.callbackId;
}

- (void)setRewardCenterClosedCallback:(CDVInvokedUrlCommand*)command
{
    onRewardCenterClosedCallback = command.callbackId;
}

- (void)setTheoremreachSurveyAvailableCallback:(CDVInvokedUrlCommand*)command
{
    theoremreachSurveyAvailableCallback = command.callbackId;
}

- (void)onReward: (NSNumber* )quantity {
    // award user the content!
    NSLog(@"TheoremReachPlugin onReward: %@", quantity);
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsNSInteger:[quantity integerValue]];
    pluginResult.keepCallback = @1;
    [self.commandDelegate sendPluginResult:pluginResult callbackId:onRewardCallback];
}

- (void)onRewardCenterOpened {
    // reward center opened! Time to take surveys!
    NSLog(@"TheoremReachPlugin onRewardCenterOpened");
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    pluginResult.keepCallback = @1;
    [self.commandDelegate sendPluginResult:pluginResult callbackId:onRewardCenterOpenedCallback];
}

- (void)onRewardCenterClosed {
    // reward center opened! Back to the app to use our coins!
    NSLog(@"TheoremReachPlugin onRewardCenterClosed");
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    pluginResult.keepCallback = @1;
    [self.commandDelegate sendPluginResult:pluginResult callbackId:onRewardCenterClosedCallback];
}

-(void)theoremreachSurveyAvailable: (BOOL) surveyAvailable {
    // TheoremReachSurvayAvailable
    NSLog(@"TheoremReachSurvayAvailable: %d", surveyAvailable);
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:surveyAvailable];
    pluginResult.keepCallback = @1;
    [self.commandDelegate sendPluginResult:pluginResult callbackId:theoremreachSurveyAvailableCallback];
}

@end

File: www/theoremreachplugin.js
var exec = require('cordova/exec');

var TheoremReachPlugin = function() {
    var onRewardCallback = function(quantity) {};
    var onRewardCenterOpenedCallback = function() {};
    var onRewardCenterClosedCallback = function() {};
    var theoremreachSurveyAvailableCallback = function(surveyAvailable) {};
}

// INIT THEOREMREACH //
TheoremReachPlugin.prototype.initWithApiKeyAndUserId = function(key, user) {
    exec(TheoremReachPlugin.onRewardCallback, function() {}, "TheoremReachPlugin", 'setRewardCallback', []);
    exec(TheoremReachPlugin.onRewardCenterOpenedCallback, function() {}, "TheoremReachPlugin", 'setRewardCenterOpenedCallback', []);
    exec(TheoremReachPlugin.onRewardCenterClosedCallback, function() {}, "TheoremReachPlugin", 'setRewardCenterClosedCallback', []);
    exec(TheoremReachPlugin.theoremreachSurveyAvailableCallback, function() {}, "TheoremReachPlugin", 'setRewardCenterClosedCallback', []);

    exec(function() {}, function() {}, "TheoremReachPlugin", 'initWithApiKeyAndUserId', [key, user]);
}

// SETTING THEOREMREACH DEBUG MODE //
TheoremReachPlugin.prototype.enableDebugMode = function(mode) {
    exec(function() {}, function() {}, "TheoremReachPlugin", 'enableDebugMode', [mode]);
}

// CHECK SURVEY AVAILABLE //
TheoremReachPlugin.prototype.isSurveyAvailable = function(success) {
    exec(success, function() {}, "TheoremReachPlugin", 'isSurveyAvailable', []);
}

// SHOW REWARD CENTER //
TheoremReachPlugin.prototype.showRewardCenter = function() {
    exec(function() {}, function() {}, "TheoremReachPlugin", 'showRewardCenter', []);
}

// REWARD CALLBACK //
TheoremReachPlugin.prototype.onReward = function(callback) {
    TheoremReachPlugin.onRewardCallback = callback;
    exec(TheoremReachPlugin.onRewardCallback, function() {}, "TheoremReachPlugin", 'setRewardCallback', []);
}

// REWARDCENTER OPENED CALLBACK //
TheoremReachPlugin.prototype.onRewardCenterOpened = function(callback) {
    TheoremReachPlugin.onRewardCenterOpenedCallback = callback;
    exec(TheoremReachPlugin.onRewardCenterOpenedCallback, function() {}, "TheoremReachPlugin", 'setRewardCenterOpenedCallback', []);
}

// REWARDCENTER CLOSED CALLBACK //
TheoremReachPlugin.prototype.onRewardCenterClosed = function(callback) {
    TheoremReachPlugin.onRewardCenterClosedCallback = callback;
    exec(TheoremReachPlugin.onRewardCenterClosedCallback, function() {}, "TheoremReachPlugin", 'setRewardCenterClosedCallback', []);
}

// THEOREMREACH AVAILABLE CALLBACK //
TheoremReachPlugin.prototype.theoremreachSurveyAvailable = function(callback) {
    TheoremReachPlugin.theoremreachSurveyAvailableCallback = callback;
    exec(TheoremReachPlugin.theoremreachSurveyAvailableCallback, function() {}, "TheoremReachPlugin", 'setTheoremreachSurveyAvailableCallback', []);
}

var theoremReachPlugin = new TheoremReachPlugin();
module.exports = theoremReachPlugin;